components/raftstore/src/store/fsm/apply.rs

```diff
@@ -2981,6 +3045,36 @@ where
             ));
         }
 
+        let voter_replicated_index = req.get_compact_log().get_voter_replicated_index();
+        // If there is any voter lagging behind, the log truncation of the witness
+        // shouldn't be triggered even if it's force mode(raft log size/count exceeds
+        // the threshold or raft engine purge), otherwise the witness can't help the
+        // lagging voter catch up logs when leader is down. In this situation Compact
+        // index should be queued. If witness receives a voter_replicated_index
+        // that is larger than the pending compact index, logs can be deleted.
```

Comment:
Question: why don't we change the logic in on_raft_gc_log_tick to make sure the correct compact_index is used under witness & lag peer case. Other than patching the logic here.

Reply:
Adjusting compact_index may cause some problems, such as not being able to GC in time and causing OOM, etc.